### ISSUE TITLE
Inline block semantics impacts text wrapping

### ISSUE THREAD

**Issue by nigelmegitt** [w3c/ttml2#146]
[GA] Add dependency
- [x] #296

Where inline block semantics apply, which give good and desirable effects in most cases, there's one extremely undesirable effect, which is that those inline blocks are treated as unbreakable units for line wrapping in the surrounding block.

Consider:

``` xml
<style xml:id="s1" tts:backgroundColor="black" tts:color="white" tts:textAlign="left">
...
<p tts:textAlign="center">
<span style="s1">This line would normally not wrap but the author put too many words in so it does</span>
<br/>
<span style="s1">This line is short</span>
```

which results in more than one line area for the first long line, with a single background rectangle covering all of the line areas, whereas the desired behaviour was to apply the background only to the each individual line area.

A CSS equivalent without `display: inline-block;` on the span renders on Chrome as:

![e1 normal](https://cloud.githubusercontent.com/assets/4988915/12954660/695e1dce-d017-11e5-8315-25a55ef1c04c.png)

whereas with `display: inline-block;` it becomes:

![e1 inline block](https://cloud.githubusercontent.com/assets/4988915/12954680/7eec629a-d017-11e5-9e99-b12f46068be1.png)

Or this:

``` xml
<p>
<span tts:bpd="auto" begin="0s">A</span><span tts:bpd="auto" begin="1s"> new</span><span tts:bpd="auto" begin="2s"> word.</span><span tts:bpd="auto" begin="3s"> A whole sentence.</span>
</p>
```

In this case each word is in its own span to apply timing, and coincidentally we also want to apply some styling that switches on inline block semantics. Now we get three unwanted effects:
1. the blank space at the beginning of each span, which would have been displayed previously is now stripped out as blank space unless we also take care to preserve it.
2. if we preserve the blank space and the span wraps onto a new line then it appears at the beginning of the new line instead of being stripped off as we would expect with normal wrapping.
3. the " A whole sentence." span can not be broken at all, even if it should be - it wraps as a single unit.

Rendered CSS using Chrome without `display: inline-block;` gives:

![e2 normal](https://cloud.githubusercontent.com/assets/4988915/12954744/c262bb64-d017-11e5-890a-d32449aa2a15.png)

and setting `display: inline-block;` on the spans it becomes:

![e2 inline block](https://cloud.githubusercontent.com/assets/4988915/12954759/d4f95a12-d017-11e5-8eda-3d2f57f43d4c.png)

We need to rethink the use of inline block semantics especially with respect to wrapping. Would it be feasible to state that inline block semantics apply to the post-layout inline areas only?

**Comment by nigelmegitt** [comment]
I've added some images to the issue description to illustrate the points more clearly. 

@skynavga I see you've changed the issue title from '... breaks text wrapping' to '... impacts text wrapping': for me the impact is that it's broken! The issue name isn't a big deal obviously as long as it's clear what we're talking about.

**Comment by skynavga** [comment]
so, (1) you are mixing a number of issues here, specifically line breaking, whitespace handling, and background painting; (3) you could fix your first example above simply by using an inner span for the background color and leave the outer span with text align only; (4) in the rather contrived second example, you could easily fix by inserting spaces in places where they have significance or using NBSP;

in conclusion, nothing is broken if it works as defined: caveat emptor applies

**Comment by nigelmegitt** [comment]
Thanks for the tips - the nested span approach does indeed work: I hadn't thought of doing that. 

The general point I'm making is that there's an interaction between line breaking, whitespace handling and background painting that seems hard to avoid, i.e. they're not orthogonal in TTML.

Did you have a point number 2?

The second example is not contrived at all - it's exactly the sort of output that we will be generating for live word by word subtitles. Inserting spaces or NBSP doesn't necessarily help because if the inline-block span containing an initial space is wrapped to the beginning of a new line area then the desired behaviour is to remove that initial space not to show it. I'll do some more investigations based on nested spans to see if I can work out how to resolve that also.

**Comment by skynavga** [comment]
I say contrived because of the use of tts:bpd='auto' to force each word span to be treated as an inline-block; why would one do this? and if doing this knowingly, why not insert white space between those word spans?

**Comment by nigelmegitt** [comment]
Ah right, I see. That rather begs the question why one would use tts:bpd at all, which could do with some exemplar explanation as already highlighted by the ed notes.

I added it simply as a way to indicate how, by setting a style attribute that might be expected to have no impact on white space handling and line wrapping (or indeed any effect at all since "auto" is the default behaviour when the style attribute is not set at all), in fact there is an impact. I'm happy to wait to see what the examples and explanation are for bpd before claiming that a change is needed.

Regardless, it would be useful minimally to include some explanatory text around inline block semantics that highlights that they have this effect and that a way to avoid it is to use nested spans. Happy to contribute these examples to that text.

**Comment by nigelmegitt** [comment]
[Meeting 2017-01-13] Agreed to separate inline-block semantics from ipd, bpd etc. and put it into the `tts:display` attribute as a new available value for use only on `span`.

**Comment by skynavga** [comment]
Add dependency
- [ ] #296

**Comment by skynavga** [comment]
It seems like the original point of this issue, namely, that inline block semantics are unusual, is non-actionable. Note that we have already removed the automatic semantics for upgrading a span to inline block, which now must be done explicitly with ``tts:display="inlineBlock"``, about which see #297.

Therefore, I am closing this as works for me. If someone wants to propose some informative language to add as one or more notes regarding inline block behavior, then feel free to post a PR and re-open this issue.

**Comment by nigelmegitt** [comment]
Thanks @skynavga I think making inlineBlock semantics require an explicit setting means this works for me too.
